group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Timer.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.runtime;
+
+import org.apache.kafka.common.annotation.InterfaceStability;
+
+/**
+ * An interface to schedule and cancel operations.
+ */
+@InterfaceStability.Unstable
+public interface Timer {
+
+    /**
+     * Add an operation to the timer. If an operation with the same key
+     * already exists, replace it with the new operation.
+     *
+     * @param key         The key to identify this operation.
+     * @param deadlineMs  The deadline to expire the operation in milliseconds.
+     * @param operation   The operation to perform.
+     */
+    void schedule(String key, long deadlineMs, Runnable operation);
```

Comment:
As a second though, I also wonder if we should use a `delay` instead of a deadline. This may be easier to use. We could also use `TimeUnit`.